asset-transfer-private-data/chaincode-go/go/privateAssetTransfer.go

```diff
@@ -0,0 +1,503 @@
+	}
+	if assetInput.Size <= 0 {
+		return fmt.Errorf("size field must be a positive integer")
+	}
+	if assetInput.AppraisedValue <= 0 {
+		return fmt.Errorf("AppraisedValue field must be a positive integer")
+	}
+
+	// Check if asset already exists
+	assetAsBytes, err := ctx.GetStub().GetPrivateData("assetCollection", assetInput.ID)
+	if err != nil {
+		return fmt.Errorf("Failed to get asset: " + err.Error())
+	} else if assetAsBytes != nil {
+		fmt.Println("This asset already exists: " + assetInput.ID)
+		return fmt.Errorf("This asset already exists: " + assetInput.ID)
+	}
+
+	// Get ID of submitting client identity
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+
+	// Make submitting client the owner
+	asset := &Asset{
+		ObjectType: assetInput.ObjectType,
+		ID:       	assetInput.ID,
+		Color:      assetInput.Color,
+		Size:       assetInput.Size,
+		Owner:      clientID,
+	}
+	assetJSONasBytes, err := json.Marshal(asset)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	// Save asset to private data collection
+	err = ctx.GetStub().PutPrivateData("assetCollection", assetInput.ID, assetJSONasBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put Asset: %s", err.Error())
+	}
+
+	// Save asset details to collection visible to owning organization
+	assetPrivateDetails := &AssetPrivateDetails{
+		ID:       			 	assetInput.ID,
+		AppraisedValue:   assetInput.AppraisedValue,
+	}
+
+	assetPrivateDetailsAsBytes, err := json.Marshal(assetPrivateDetails) // marshal asset details to JSON
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	// get collection name for this organization
+	collection, err := getCollectionName(ctx)
+
+	// put asset into owners private data collection
+	err = ctx.GetStub().PutPrivateData(collection, assetInput.ID, assetPrivateDetailsAsBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put asset private details: %s", err.Error())
+	}
+
+	//  ==== Index the asset to enable color-based range queries, e.g. return all blue assets ====
+	//  An 'index' is a normal key/value entry in state.
+	//  The key is a composite key, with the elements that you want to range query on listed first.
+	//  In our case, the composite key is based on indexID=color~ID.
+	//  This will enable very efficient state range queries based on composite keys matching indexID=color~*
+	indexID := "color~ID"
+	colorIDIndexKey, err := ctx.GetStub().CreateCompositeKey(indexID, []string{asset.Color, asset.ID})
+	if err != nil {
+			return err
+	}
+	//  Save index entry to state. Only the key ID is needed, no need to store a duplicate copy of the asset.
+	//  Note - passing a 'nil' value will effectively delete the key from state, therefore we pass null character as value
+	value := []byte{0x00}
+	err = ctx.GetStub().PutPrivateData("assetCollection", colorIDIndexKey, value)
+
+	// ==== Asset saved and indexed. Return success ====
```

Comment:
This came from the original marbles02, but can be deleted at this point since the private chaincode doesn't transfer by color anymore.